segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/SystemJournal.java

```diff
@@ -0,0 +1,483 @@
+/**
+ * Copyright (c) Dell Inc., or its subsidiaries. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.pravega.segmentstore.storage.chunklayer;
+
+import com.google.common.base.Preconditions;
+import io.pravega.segmentstore.storage.SegmentRollingPolicy;
+import io.pravega.segmentstore.storage.metadata.ChunkMetadata;
+import io.pravega.segmentstore.storage.metadata.ChunkMetadataStore;
+import io.pravega.segmentstore.storage.metadata.MetadataTransaction;
+import io.pravega.segmentstore.storage.metadata.SegmentMetadata;
+import io.pravega.segmentstore.storage.metadata.StorageMetadataException;
+import io.pravega.shared.NameUtils;
+import lombok.Getter;
+import lombok.Setter;
+import lombok.extern.slf4j.Slf4j;
+import lombok.val;
+import java.io.BufferedReader;
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.HashMap;
+
+/**
+ * This class implements system journaling functionality for critical storage system segments which is useful for bootstrap after failover.
+ * It records any layout changes to storage system segments.
+ * Storage system segments are the segments that the storage subsystem uses to store all metadata.
+ * This create a circular dependency while reading or writing the data about these segment from the metadata segments..
+ * System journal is a mechanism to break this circular dependency by having independent log of all layout changes to system segments.
+ * Currently only two actions are considered viz. Addition of new chunks and truncation of segments.
+ * This log is replayed when the ChunkStorageManager is booted.
+ * To avoid data corruption. Each instance writes to its own distinct log file.
+ * During bootstrap all the system journal files are read and processed to re-create the state of the storage system segments.
+ */
+@Slf4j
+public class SystemJournal {
+    private static final String ADD_RECORD = "ADD";
```

Comment:
After reading this file, I believe it will will benefit from serialization versioning. Consider using `VersionedSerializer`, couple with binary encoding to both reduce the size of your serializations and future-proof them from change.

Reply:
At the current moment I need debug friendly format. 
Once I am done with all changes, I do plan to implement serialization at the very end as part of the last PR in this group.